File: src/v2/chromatic_circle/ChromaticCircle.ts
import { ColorMixer } from "./mixers/ColorMixer";
import { Color } from "./Color";
import { BlueMixer } from "./mixers/BlueMixer";
import { RedMixer } from "./mixers/RedMixer";
import { YellowMixer } from "./mixers/YellowMixer";

export class ChromaticCircle {
  private readonly filter: Color;
  private readonly mixers: ColorMixer[];

  constructor(filter: Color) {
    this.filter = filter;
    this.mixers = [new BlueMixer(), new RedMixer(), new YellowMixer()];
  }

  public mixColors(color: Color): Color {
    const colorMixer = this.mixers.find((mixer) => mixer.canMix(this.filter));
    if (!colorMixer) {
      throw new Error(`No color mixer found that can handle filter: ${this.filter}`);
    }
    return colorMixer.mixColors(color);
  }
}

File: java/src/main/java/com/circulocromatico/ChromaticCircle.java
package com.circulocromatico;

import java.util.Arrays;
import java.util.stream.Stream;

/**
 * ChromaticCircle class that simulates a color mixing system based on a filter.
 * The filter can be one of the primary colors: RED, BLUE, or YELLOW.
 * It mixes colors according to the rules of color theory.
 */
public class ChromaticCircle {
    private final Color filter;
    private final ColorMixer[] mixers;

    public ChromaticCircle(Color filter) {
        this.filter = filter;
        ColorMixer blueMixer = new BlueMixer();
        ColorMixer colorMixer = new RedMixer();
        YellowMixer yellowMixer = new YellowMixer();
        this.mixers = new ColorMixer[] {blueMixer, colorMixer, yellowMixer};
    }

    public Color mixColors(Color color) {
        return Arrays.stream(mixers).dropWhile(mixer -> !mixer.canMix(filter))
                .findFirst()
                .orElseThrow(() -> new IllegalArgumentException("No mixer found for filter: " + filter))
                .mixColor(color);
    }

    public enum Color {
        RED,
        BLUE,
        YELLOW,
        GREEN,
        ORANGE,
        VIOLET
    }
}
